common/lib/plugins/efm2/monitor.ts

```diff
@@ -0,0 +1,310 @@
+            while ((monitorContextRef = queue.shift()) != null) {
+              const monitorContext: MonitorConnectionContext = monitorContextRef.deref();
+              if (monitorContext !== null && monitorContext.isActive()) {
+                this.activeContexts.push(monitorContextRef);
+              }
+            }
+          }
+        }
+        processedKeys.forEach((key) => {
+          MonitorImpl.newContexts.delete(key);
+        });
+        await sleep(1000);
+      }
+      return;
+    } catch (err) {
+      // do nothing, exit task
+    }
+    logger.debug(Messages.get("MonitorImpl.stopMonitoringTaskNewContext", this.hostInfo.host));
+  }
+
+  async run(): Promise<void> {
+    logger.debug(Messages.get("MonitorImpl.startMonitoring", this.hostInfo.host));
+
+    try {
+      while (!this.isStopped()) {
+        try {
+          if (this.activeContexts.length === 0 && !this.hostUnhealthy) {
+            await sleep(MonitorImpl.TASK_SLEEP_NANOS / 1000000);
+          }
+
+          const statusCheckStartTimeNanos: number = this.getCurrentTimeNano();
+          const isValid = await this.checkConnectionStatus();
+          const statusCheckEndTimeNanos: number = this.getCurrentTimeNano();
+
+          await this.updateHostHealthStatus(isValid, statusCheckStartTimeNanos, statusCheckEndTimeNanos);
+
+          if (this.hostUnhealthy) {
+            this.pluginService.setAvailability(this.hostInfo.aliases, HostAvailability.NOT_AVAILABLE);
+          }
+          const tmpActiveContexts: WeakRef<MonitorConnectionContext>[] = [];
+
+          let monitorContextRef: WeakRef<MonitorConnectionContext> | undefined;
+
+          while ((monitorContextRef = this.activeContexts.shift()) != null) {
+            if (this.isStopped()) {
+              break;
+            }
+            const monitorContext = monitorContextRef.deref();
+            if (monitorContext == null) {
+              continue;
+            }
+
+            if (this.hostUnhealthy) {
+              // Kill connection
+              monitorContext.setHostUnhealthy(true);
+              monitorContext.setInactive();
+              const connectionToAbort = monitorContext.getClient();
+              if (connectionToAbort != null) {
+                await this.endMonitoringClient();
+                this.abortedConnectionsCounter.inc();
+              }
+            } else if (monitorContext.isActive()) {
+              tmpActiveContexts.push(monitorContextRef);
+            }
+
+            // activeContexts is empty now and tmpActiveContexts contains all yet active contexts
+            // Add active contexts back to the queue.
+            this.activeContexts.push(...tmpActiveContexts);
+
+            const delayNanos = this.failureDetectionIntervalNanos - (statusCheckEndTimeNanos - statusCheckStartTimeNanos);
+            await sleep(Math.max(delayNanos, MonitorImpl.TASK_SLEEP_NANOS) / 1000000);
```

Comment:
It might be helpful to call `clearTimeout` on potentially longer sleep statements like this when releasing resources